docs/esbulk.1
.TH ESBULK 1 "JANUAR 2018" "Leipzig University Library" "Manuals"
.SH NAME
.PP
esbulk \- send documents to elasticsearch in bulk and in parallel.

.SH SYNOPSIS
.PP
\fB\fCesbulk\fR [\fB\fC\-server\fR \fIURL\fP, \fB\fC\-index\fR \fIname\fP, \fB\fC\-size\fR \fIN\fP, \fB\fC\-w\fR \fIN\fP, \fB\fC\-z\fR] < \fIfile\fP

.SH DESCRIPTION
.PP
esbulk takes as input a newline delimited JSON file and indexes all documents
into elasticsearch running on a given server address. The documents are batched
and indexed in parallel to achieve a high indexing throughput.

.PP
The newline delimited JSON text file format is explained at 
\[la]http://jsonlines.org/\[ra] and 
\[la]http://ndjson.org/\[ra]\&.

.SH OPTIONS
.PP
\fB\fC\-0\fR
  Set the number of replicas to 0 during indexing (this can speed up indexing significantly, the original value is restored at the end and may cause delay until the cluster is green).

.PP
\fB\fC\-apikey\fR \fIstring\fP
  Set the encoded ES api key (mutually exclusive with -u).


.PP
\fB\fC\-c\fR \fIstring\fP
  Create index mappings, settings, aliases, 
\[la]https://is.gd/3zszeu\[ra]\&.

.PP
\fB\fC\-cpuprofile\fR \fIstring\fP
  Write cpu profile to file.

.PP
\fB\fC\-id\fR \fIstring\fP
  Reuse value from this field as id. By default ids are autogenerated.

.PP
\fB\fC\-index\fR \fIstring\fP
  Index name.

.PP
\fB\fC\-mapping\fR \fIfilename\fP
  Mapping string or filename to apply before indexing.

.PP
\fB\fC\-memprofile\fR \fIstring\fP
  Write heap profile to file.

.PP
\fB\fC\-optype\fR \fIstring\fP
  optype (index \- will replace existing data, create \- will only create a new doc,
  update \- create new or update existing data) (default "index")

.PP
\fB\fC\-p\fR \fIname\fP
  Pipeline to use to preprocess documents.

.PP
\fB\fC\-purge\fR
  Purge any existing index before reindexing. Warning: No confirmation required.

.PP
\fB\fC\-r string\fR
  Refresh interval after import (default "1s")

.PP
\fB\fC\-server\fR \fIURL\fP
  Server hostport including schema like 
\[la]http://localhost:9200\[ra]

.PP
\fB\fC\-size\fR \fIN\fP
  Batch size. Defaults to 1000. Increase for small documents.

.PP
\fB\fC\-skipbroken\fR
  Skip broken json.

.PP
\fB\fC\-type\fR \fIstring\fP
  Elasticsearch type (deprecated in 6.0.0, 
\[la]https://is.gd/HFsOWt\[ra]), empty string.

.PP
\fB\fC\-u\fR \fIstring\fP
  HTTP basic authentication "username:password" (like curl \-u).

.PP
\fB\fC\-v\fR
  Program version.

.PP
\fB\fC\-verbose\fR
  Show progress.

.PP
\fB\fC\-w\fR \fIN\fP
  Number of workers. Defaults to number of cores.

.PP
\fB\fC\-z\fR
  Decompress gzip input file on the fly.

.SH EXAMPLES
.PP
Index a compressed file:

.PP
\fB\fCesbulk \-index abc \-verbose \-server 110.81.131.200:9200 \-z file.ldj.gz\fR

.PP
Index from standard input:

.PP
\fB\fCcat file.ldj | esbulk \-index abc \-server 110.81.131.200:9200\fR

.PP
Purge an existing index, apply a mapping from a file and index:

.PP
\fB\fCesbulk \-purge \-mapping mapping.json \-index abc file.ldj\fR

.SH DIAGNOSITCS
.PP
If indexing pressure on the bulk API is too high (dozens or hundreds of
parallel workers, large batch sizes, depending on you setup), esbulk will halt
and report an error:

.PP
.RS

.nf
$ esbulk \-index my\-index\-name \-w 100 file.ldj
2017/01/02 16:25:25 error during bulk operation, try less workers (lower \-w value) or
increase thread\_pool.bulk.queue\_size in your nodes

.fi
.RE

.PP
Please note that, in such a case, some documents are indexed and some are not.
Your index will be in an INCONSISTENT STATE, since there is no transactional
bracket around the indexing process.

.PP
However, using defaults (parallism: number of cores) on a single node setup
will just work. For larger clusters, increase the number of workers until you
see full CPU utilization. After that, more workers won't buy any more speed.

.SH BUGS
.PP
Please report bugs to 
\[la]https://github.com/miku/esbulk/issues\[ra]\&.

.SH AUTHORS
.PP
Martin Czygan 
\[la]https://github.com/miku\[ra], 
\[la]martin.czygan@uni-leipzig.de\[ra]
sakshambathla 
\[la]https://github.com/sakshambathla\[ra]
Klaubert Herr 
\[la]https://github.com/klaubert\[ra]
Yusuke KUOKA 
\[la]https://github.com/mumoshu\[ra]
faultlin3 
\[la]https://github.com/faultlin3\[ra]
gransy 
\[la]https://github.com/gransy\[ra]
Christoph Kepper 
\[la]https://github.com/ckepper\[ra]
Christian Solomon
Mikael Byström

.SH SEE ALSO
.PP
FINC
\[la]https://finc.info\[ra], AMSL
\[la]http://amsl.technology/\[ra], solrbulk
\[la]https://github.com/miku/solrbulk\[ra]


cmd/esbulk/esbulk.go
package main

import (
	"bufio"
	"compress/gzip"
	"encoding/json"
	"flag"
	"fmt"
	"io"
	"log"
	"math/rand"
	"net/http"
	"os"
	"runtime"
	"runtime/pprof"
	"strings"
	"sync"
	"time"

	"github.com/miku/esbulk"
	"github.com/sethgrid/pester"
)

// Version of application.
const Version = "0.6.2"

var (
	version         = flag.Bool("v", false, "prints current program version")
	cpuprofile      = flag.String("cpuprofile", "", "write cpu profile to file")
	memprofile      = flag.String("memprofile", "", "write heap profile to file")
	indexName       = flag.String("index", "", "index name")
	docType         = flag.String("type", "default", "elasticsearch doc type")
	batchSize       = flag.Int("size", 1000, "bulk batch size")
	numWorkers      = flag.Int("w", runtime.NumCPU(), "number of workers to use")
	verbose         = flag.Bool("verbose", false, "output basic progress")
	skipbroken      = flag.Bool("skipbroken", false, "skip broken json")
	gzipped         = flag.Bool("z", false, "unzip gz'd file on the fly")
	mapping         = flag.String("mapping", "", "mapping string or filename to apply before indexing")
	purge           = flag.Bool("purge", false, "purge any existing index before indexing")
	idfield         = flag.String("id", "", "name of field to use as id field, by default ids are autogenerated")
	user            = flag.String("u", "", "http basic auth username:password, like curl -u")
	zeroReplica     = flag.Bool("0", false, "set the number of replicas to 0 during indexing")
	refreshInterval = flag.String("r", "1s", "Refresh interval after import")
	pipeline        = flag.String("p", "", "pipeline to use to preprocess documents")
)

//isJSON checks if a string is valid json
func IsJSON(str string) bool {
	var js json.RawMessage
	return json.Unmarshal([]byte(str), &js) == nil
}

// indexSettingsRequest runs updates an index setting, given a body and
// options. Body consist of the JSON document, e.g. `{"index":
// {"refresh_interval": "1s"}}`.
func indexSettingsRequest(body string, options esbulk.Options) (*http.Response, error) {
	r := strings.NewReader(body)

	rand.Seed(time.Now().Unix())
	server := options.Servers[rand.Intn(len(options.Servers))]
	link := fmt.Sprintf("%s/%s/_settings", server, options.Index)

	req, err := http.NewRequest("PUT", link, r)
	if err != nil {
		return nil, err
	}
	// Auth handling.
	if options.Username != "" && options.Password != "" {
		req.SetBasicAuth(options.Username, options.Password)
	}
	req.Header.Set("Content-Type", "application/json")

	resp, err := pester.Do(req)
	if err != nil {
		return nil, err
	}
	if options.Verbose {
		log.Printf("applied setting: %s with status %s\n", body, resp.Status)
	}
	return resp, nil
}

func main() {
	var serverFlags esbulk.ArrayFlags
	flag.Var(&serverFlags, "server", "elasticsearch server, this works with https as well")

	flag.Parse()

	if *cpuprofile != "" {
		f, err := os.Create(*cpuprofile)
		if err != nil {
			log.Fatal(err)
		}
		pprof.StartCPUProfile(f)
		defer pprof.StopCPUProfile()
	}

	if *version {
		fmt.Println(Version)
		os.Exit(0)
	}

	if *indexName == "" {
		log.Fatal("index name required")
	}

	if len(serverFlags) == 0 {
		serverFlags = append(serverFlags, "http://localhost:9200")
	}

	if *verbose {
		log.Printf("using %d servers", len(serverFlags))
	}

	var file io.Reader = os.Stdin

	if flag.NArg() > 0 {
		f, err := os.Open(flag.Arg(0))
		if err != nil {
			log.Fatalln(err)
		}
		defer f.Close()
		file = f
	}

	runtime.GOMAXPROCS(*numWorkers)

	var username, password string
	if len(*user) > 0 {
		parts := strings.Split(*user, ":")
		if len(parts) != 2 {
			log.Fatal("http basic auth syntax is: username:password")
		}
		username = parts[0]
		password = parts[1]
	}

	options := esbulk.Options{
		Servers:   serverFlags,
		Index:     *indexName,
		DocType:   *docType,
		BatchSize: *batchSize,
		Verbose:   *verbose,
		Scheme:    "http",
		IDField:   *idfield,
		Username:  username,
		Password:  password,
		Pipeline:  *pipeline,
	}

	if *verbose {
		log.Println(options)
	}

	if *purge {
		if err := esbulk.DeleteIndex(options); err != nil {
			log.Fatal(err)
		}
		time.Sleep(5 * time.Second)
	}

	// Create index if not exists.
	if err := esbulk.CreateIndex(options); err != nil {
		log.Fatal(err)
	}

	if *mapping != "" {
		var reader io.Reader
		if _, err := os.Stat(*mapping); os.IsNotExist(err) {
			reader = strings.NewReader(*mapping)
		} else {
			file, err := os.Open(*mapping)
			if err != nil {
				log.Fatal(err)
			}
			reader = bufio.NewReader(file)
		}
		err := esbulk.PutMapping(options, reader)
		if err != nil {
			log.Fatal(err)
		}
	}

	queue := make(chan string)
	var wg sync.WaitGroup

	for i := 0; i < *numWorkers; i++ {
		wg.Add(1)
		go esbulk.Worker(fmt.Sprintf("worker-%d", i), options, queue, &wg)
	}

	for i, _ := range options.Servers {
		// Store number_of_replicas settings for restoration later.
		doc, err := esbulk.GetSettings(i, options)
		if err != nil {
			log.Fatal(err)
		}

		// TODO(miku): Rework this.
		numberOfReplicas := doc[options.Index].(map[string]interface{})["settings"].(map[string]interface{})["index"].(map[string]interface{})["number_of_replicas"]
		if *verbose {
			log.Printf("on shutdown, number_of_replicas will be set back to %s", numberOfReplicas)
		}
		if *verbose {
			log.Printf("on shutdown, refresh_interval will be set back to %s", *refreshInterval)
		}
		// Shutdown procedure. TODO(miku): Handle signals, too.
		defer func() {
			// Realtime search.
			if _, err := indexSettingsRequest(fmt.Sprintf(`{"index": {"refresh_interval": "%s"}}`, *refreshInterval), options); err != nil {
				log.Fatal(err)
			}
			// Reset number of replicas.
			if _, err := indexSettingsRequest(fmt.Sprintf(`{"index": {"number_of_replicas": %q}}`, numberOfReplicas), options); err != nil {
				log.Fatal(err)
			}
			// Persist documents.
			if err := esbulk.FlushIndex(i, options); err != nil {
				log.Fatal(err)
			}
		}()

		// Realtime search.
		resp, err := indexSettingsRequest(`{"index": {"refresh_interval": "-1"}}`, options)
		if err != nil {
			log.Fatal(err)
		}
		if resp.StatusCode >= 400 {
			log.Fatal(resp)
		}
		if *zeroReplica {
			// Reset number of replicas.
			if _, err := indexSettingsRequest(`{"index": {"number_of_replicas": 0}}`, options); err != nil {
				log.Fatal(err)
			}
		}
	}

	reader := bufio.NewReader(file)
	if *gzipped {
		zreader, err := gzip.NewReader(file)
		if err != nil {
			log.Fatal(err)
		}
		reader = bufio.NewReader(zreader)
	}

	counter := 0
	start := time.Now()

	for {
		line, err := reader.ReadString('\n')
		if err == io.EOF {
			break
		}
		if err != nil {
			log.Fatal(err)
		}
		line = strings.TrimSpace(line)
		if len(line) == 0 {
			continue
		}
		if *skipbroken {
			if !(IsJSON(line)) {
				if *verbose {
					fmt.Printf("Skipped line [%s]\n", line)
				}
				continue
			}
		}

		queue <- line
		counter++
	}

	close(queue)
	wg.Wait()
	elapsed := time.Since(start)

	if *memprofile != "" {
		f, err := os.Create(*memprofile)
		if err != nil {
			log.Fatal(err)
		}
		pprof.WriteHeapProfile(f)
		f.Close()
	}

	if *verbose {
		rate := float64(counter) / elapsed.Seconds()
		log.Printf("%d docs in %s at %0.3f docs/s with %d workers\n", counter, elapsed, rate, *numWorkers)
	}
}
